### ISSUE TITLE
The current testing environment is not configured to support act(...) with vitest and React 18

### ISSUE THREAD

**Issue by MarkLyck** [testing-library/react-testing-library#1061]
"@testing-library/jest-dom": "5.16.4",
"@testing-library/react": "13.1.1",
DOM Environment: jsdom or happy-dom (happens with both)

### Problem description:

After upgrading to @testing-library/react@13.x and react@18.x with `vitest`. A lot of react tests throw the following warning:

```
Warning: The current testing environment is not configured to support act(...)
```

I noticed this after switching from jest to vitest, and not changing anything else. (The tests do not show this warning when using jest, but they they show it when using vitest in the same project).

I previously had an unrelated issue causing the same warning with Jest: https://github.com/testing-library/react-testing-library/issues/1057

I resolved that, by removing a bad implementation of `act()`. However this time the tests are identical, and works with 1 test-runner and not with another.

### Reproduction:
Credit to @[stuarthallows](https://github.com/stuarthallows) for creating a repo that reproduces the problem: https://github.com/stuarthallows/vitest-error

@stuarthallows and @dar5hak also both seem to be having the same problem as me (see my other issue here: https://github.com/testing-library/react-testing-library/issues/1057 for more details).

### Additional information.

In my case, this is breaking unit tests completely because it logs thousands of these warnings and it just stops after around 10 tests are completed.

For me a lot of the errors start with `rc-motion`

```
Warning: The current testing environment is not configured to support act(...)
    at Align (/Users/marklyck/colony/colony-frontend/node_modules/.pnpm/rc-align@4.0.12_ef5jwxihqo6n7gxfmzogljlgcm/node_modules/rc-align/lib/Align.js:45:23)
    at DomWrapper (/Users/marklyck/colony/colony-frontend/node_modules/.pnpm/rc-motion@2.5.1_ef5jwxihqo6n7gxfmzogljlgcm/node_modules/rc-motion/lib/DomWrapper.js:28:34)
    at /Users/marklyck/colony/colony-frontend/node_modules/.pnpm/rc-motion@2.5.1_ef5jwxihqo6n7gxfmzogljlgcm/node_modules/rc-motion/lib/CSSMotion.js:57:32
```

Which is probably running an animation with antd components. But I'm 99% sure that's a red-herring since the same tests and components work just fine with jest. (or testing-library/react@12 and react@17)

**Comment by MarkLyck** [comment]
I found a workaround that seems to work with happy-dom (haven't tested with jsdom).

If you manually add

```
global.IS_REACT_ACT_ENVIRONMENT = true
```

To the `setupFile` the warning will go away.

This should of course be handled by `React testing library` so it's not a proper fix. But it allows us to use vitest until this issue is resolved :)

**Comment by airjp73** [comment]
The workaround doesn't totally work for me. The initial error `The current testing environment is not configured... to support act(...)` goes away, but then I get `act` warnings I wouldn't have otherwise.

I've made a reproduction here: https://github.com/airjp73/vitest-act-reproduction

Worth noting that this also appears to be exclusive to React 18. If you downgrade everything to React 17 (including downgrading this library to v12), the act warnings go away and you also don't need to override the global anymore.

**Comment by nknapp** [comment]
I raised a similar [issue at vitest](https://github.com/vitest-dev/vitest/issues/1146). But they think (and I kind of agree), that it has to be solved in testing-library.

## tldr; of my vitest-issue:

* when `testing-library/react` resolves `IS_REACT_ACT_ENVIRONMENT` it uses the first existing of `self`, `window` and `global` as global-object, which in may case is `self` (see [here](https://github.com/testing-library/react-testing-library/blob/c8c93f83228a68a270583c139972e79b1812b7d3/src/act-compat.js#L7).
* `react` access `IS_REACT_ACT_ENVIRONMENT` directly, which is probably `globalThis` or `global` ([link](https://github.com/facebook/react/blob/d5b6b4b865ebf13a1eaf2342d623101056e5e197/packages/react-reconciler/src/ReactFiberAct.old.js#L44).
* in `vitest` / `jsdom`, `globalThis` and `self` are not the same object.

## The fix 

I think the fix is to add 

```
if (typeof globalThis !== 'undefined') {
  return globalThis
}
```
here: https://github.com/testing-library/react-testing-library/blob/c8c93f83228a68a270583c139972e79b1812b7d3/src/act-compat.js#L6

## The workaround

The workaround is define some kind of proxy-property on global this, that will make sure that `self` and `globalThis` have the same value of `IS_REACT_ACT_ENVIRONMENT` the  code in `setupFile`:

```
Object.defineProperty(globalThis,"IS_REACT_ACT_ENVIRONMENT", {
  get() {
    if (typeof globalThis.self !== 'undefined') {
      return globalThis.self.IS_REACT_ACT_ENVIRONMENT
    }
  },
  set(value) {
    if (typeof globalThis.self !== 'undefined') {
      globalThis.self.IS_REACT_ACT_ENVIRONMENT = value
    }
  }
})

globalThis.IS_REACT_ACT_ENVIRONMENT = true
```

**Comment by eps1lon** [comment]
We're waiting on confirmation for https://github.com/vitest-dev/vitest/issues/1146#issuecomment-1120153302. We're matching the `globalThis` polyfill so it isn't obvious that we can or should change something.

**Comment by nknapp** [comment]
shouldn't a polyfill only apply if the feature itself is not available? I would argue that if `globalThis` exists, you should use 
it, and not the polyfill.

**Comment by eps1lon** [comment]
> shouldn't a polyfill only apply if the feature itself is not available? I would argue that if `globalThis` exists, you should use it, and not the polyfill.

We can do that. The code shouldn't necessarily be considered a polyfill. It worked perfectly well when `globalThis` didn't exist. So there's definitely something wrong in vitest environments since they wouldn't support legacy code that works in any other environment. 

We can still fix it by using `globalThis` if it exists. So feel free to send a PR. But older versions will continue to not work in vitest until they fix that issue on their side as well.

**Comment by airjp73** [comment]
I haven't had a chance to test myself, but I believe this was [fixed on vitest's end](https://github.com/vitest-dev/vitest/issues/1242).

**Comment by ulrichstark** [comment]
Yes can confirm that this issue is fixed for me with newest vitest version.

**Comment by WonderPanda** [comment]
I've upgraded to the latest version of vitest (0.12.6 at the time of writing) and it still doesn't resolve the issue. Using the workaround @nknapp [here](https://github.com/testing-library/react-testing-library/issues/1061#issuecomment-1119435440) does successfully get rid of the warnings

**Comment by stuarthallows** [comment]
Also not fixed for me with vitest 0.12.6.

**Comment by nknapp** [comment]
Thanks for creating that MR, @eps1lon. I think the issue I mentioned is indeed fixed with vitest@0.12.6

@stuarthallows I have just spent a couple of hours trying to find out why I also still get some `act`-warnings. Maybe you have a similar problem.

The following code is wrong, but the error is very hard to see

```typescript
it("sends no access token after logout", async() => {
    // ...
    await clickLogoutButton();
    await waitForLogout();
    // ...
})

async function clickLogoutButton() {
  await act(() => {
    return screen.getByText("Logout").click();
  });
}
```

There is an `await act()`, but the callback is a synchronous function. This leads to some bad "timing", so that the actual state-change is just after the `act()` call, but before the `waitFor`-call that is done inside `waitForLogout()`

The function `clickLogoutButton` should be either completely synchronous:

```typescript
it("sends no access token after logout", async() => {
    // ...
    clickLogoutButton();  // <---- no "await"
    await waitForLogout();
    // ...
})

function clickLogoutButton() {  // <--- no "async"
  act(() => {  // <--- no "await"
    screen.getByText("Logout").click();
  });
}
```

or completely asynchronous

```typescript
it("sends no access token after logout", async() => {
    // ...
    await clickLogoutButton(); 
    await waitForLogout();
    // ...
})

async function clickLogoutButton() {
  await act(async () => { // <--- added "async" to callback
    screen.getByText("Logout").click();
  });
}
``

**Comment by mindplay-dk** [comment]
> There is an `await act()`, but the callback is a synchronous function. This leads to some bad "timing", so that the actual state-change is just after the `act()` call, but before the `waitFor`-call that is done inside `waitForLogout()`

This paragraph right here was a life-saver after two solid days of fighting through a react 18 upgrade.

Apparently, if you put `find*` calls inside an `act`, the timing somehow doesn't work out... it would be reasonable to think "more awaits better" in terms of timing, but for some reason, awaiting an `act` that awaits things doesn't work out well.

The results are extremely confusing and spooky - literally your test can *pass* and then error *after* they pass!

```
$ yarn test -- pages/Home
yarn run v1.22.19
warning From Yarn 1.0 onwards, scripts don't require "--" for options to be forwarded. In a future version, any explicit "--" will be forwarded as-is to the scripts.
$ jest pages/Home
 PASS  src/pages/Home/index.test.tsx
  ● Console

    console.error
      Warning: The current testing environment is not configured to support act(...)
```

This particular project is open source, and you can see the commit that fixed my failing tests here - it has several examples of "things that don't work if you wrap them in act":

https://github.com/abtion/muffi.net/commit/b70302f1871e2382025d2f6975ec12786dbcf299

Simply removing some of the `act` wrappers fixes the timing issues.

**This should be covered by documentation maybe?**

I know you're just pulling this functionality from react's own facilities, but their documentation doesn't cover this - and I suppose maybe their documentation can't be expected to cover the interplay with facilities provided by react-testing-library?

Not sure. This was extremely confusing and time consuming and seems like something almost everyone would run into at first.

Could it be detected and guarded against, with a better error message to guide people along maybe? 🤔

**Comment by targumon** [comment]
With React 17 + @TL/react 12 + @TL/user-event 13 I had tests such as this:
```javascript
it('handles clicks', () => {
  const spy = jest.fn()
  render(<Button onClick={spy}/>)
  userEvent.click(screen.getByRole('button'))
  expect(spy).toBeCalled()
})
```
Worked just fine. I was proud of not using `act` even once in the entire project.
(btw, note my test is synchronous. async/await was only needed with more complex components)

Still working on upgrade to React 18 + @TL/react 13 + @TL/user-event 14 👷
Following the [recommendation in the docs](https://testing-library.com/docs/user-event/intro/) I switched out of `userEvent`:
```javascript
const user = userEvent.setup()
await user.click(screen.getByRole('button')) // ❗️jest throws timeout error
```
For most (simple) tests, this was fixed by adding `waitFor`:
```javascript
await waitFor(() => user.click(screen.getByRole('button')))
```

However, the more complex components would throw 'An update to Comp inside a test was not wrapped in act'.
Seems weird because I was under the impression the `waitFor` does that under the hood.
Swallowed my pride 😞 and fixed it by adding `act`:
```javascript
await act(() => waitFor(() => user.click(screen.getByRole('button')))) // not as pretty as before, but works
```

Surprisingly in _some_ cases I get 'The current testing environment is not configured to support act' (sadly I can't tell why it warns for certain components but not for others).
After hours of trying to fix it I found a solution completely by accident:
```javascript
await act(() => waitFor(() => { user.click(screen.getByRole('button')) })) // ✅ curly braces!
```

I hope this is helpful - either to the wonderful maintainers of TL or any user facing the same warning. 🙏

**Comment by PenghaiZhang** [comment]
My case is if I run my test  individually everything is good. But I run all of my tests at the same time this error comes up. 
I have `userEvent.setup({advanceTimers: jest.advanceTimersByTime})` in the test that needs some fake timing.

**Comment by binhphan009** [comment]
> 

await act(() => waitFor(() => { user.click(screen.getByRole('button')) })) // ✅ curly braces!
work for me <3

**Comment by nknapp** [comment]
Tbh. I don't think it's good practice to put actions like 'click' into a 'waitfor'.

'waitFor' is for observation. It may end up in multiple clicks if you are not careful. 
 'Waitfor' should also not be nested in 'act'. 

Those may be pragmatic workarounds, but not proper solutions.

I am writing this for everybody who is new to testing-library, reading this thread and learning from it.

If these problems still occur, it's probably a bug, in testing-library or react.

**Comment by IanTorquato** [comment]
I don't know if it helps, but I had the same error without using Vitest

```Warning: The current testing environment is not configured to support act(...)```

**Comment by targumon** [comment]
> saved my life <3

@binhphan009 You're welcome, but in fact, a couple of months after I left that comment here, I re-visited my test suites and removed almost all of the `act` calls.

It turns out I wasn't properly waiting for my API-dependent components to load (I use react-query and I obviously mock the backend in these tests, but the wait still needed to be handled) so jest timed out because the UI wasn't ready.

Wrapping `user.click` in a `waitFor` and then again in an `act` is a bad workaround (as @nknapp rightfully points out).
btw, the curly braces are ANOTHER bad workaround: they just cause the callback provided to `waitFor` to NOT return the promise from `user.click` which slightly alters its behavior in our favor, but it's just an accident, not a proper solution.

I'm now back to using `await user.click(screen.getByRole('button'))` totally error-free.

**Comment by sag1v** [comment]
> I'm now back to using await user.click(screen.getByRole('button')) totally error-free.

@targumon So what were your findings and how did you resolve it? 😄

**Comment by nikolyachikm** [comment]
had this error which randomly failed tests. in my case I had code like this (select an option in dropdown):
```
fireEvent.mouseDown(...);
fireEvent.click(...);
```
solution:
```
await waitFor(() => {
    fireEvent.mouseDown(...);
    fireEvent.click(...);
});
```

**Comment by KimSchneider** [comment]
> > I'm now back to using await user.click(screen.getByRole('button')) totally error-free.
> 
> @targumon So what were your findings and how did you resolve it? 😄

@targumon you teased an answer but I am missing code! :D

**Comment by Rita0156** [comment]
const originalError = console.error;
  beforeAll(() => {
    console.error = (...args) => {
      if (
        /Warning.*not wrapped in act/.test(args[0]) ||
        /The current testing environment is not configured to support act/.test(args[0])
      ) {
        return;
      }
      originalError.call(console, ...args);
    };
  });

  afterAll(() => {
    console.error = originalError;
  });

**Comment by Danielvandervelden** [comment]
For those of you wondering changing the following

```
describe("something", () => {
  it('should expand all options', async () => {
    render(
      <ProviderWrapper>
        <ProjectsOverview />
      </ProviderWrapper>,
    );

    const expandOptionsButton = screen.getByText('Visuals.table.actions.project.expandOptions').closest('button');
    if (!expandOptionsButton) throw new Error('Expand options Button not found');
    

    await act(async () => {
      await userEvent.click(expandOptionsButton);
    });

    expect(screen.getByText('Visuals.table.actions.option.collapseAll')).toBeInTheDocument();

    expect(screen.getByText('Wii')).toBeInTheDocument();
    expect(screen.getByText('Visuals.addReview')).toBeInTheDocument();
  });
})
```

To this:

```
describe("something", () => {
    it('should expand all options', async () => {
    render(
      <ProviderWrapper>
        <ProjectsOverview />
      </ProviderWrapper>,
    );

    const expandOptionsButton = screen.getByText('Visuals.table.actions.project.expandOptions').closest('button');
    if (!expandOptionsButton) throw new Error('Expand options Button not found');

    await userEvent.click(expandOptionsButton);

    expect(screen.getByText('Visuals.table.actions.option.collapseAll')).toBeInTheDocument();

    expect(screen.getByText('Wii')).toBeInTheDocument();
    expect(screen.getByText('Visuals.addReview')).toBeInTheDocument();
  });
})
```

Fixed my issue. A.K.A. remove the act entirely and just `await` the `userEvent.click` instead.

**Comment by MichalGorskiOnegini** [comment]
@Danielvandervelden works indeed, dear legend. Tho the reason why this happens is nicely explained here, for those who have enough time to read through it: https://stackoverflow.com/questions/75448819/async-thunks-cause-warnings-in-tests-wrapped-in-act e.g https://codilime.com/blog/why-should-you-be-grateful-for-act-warnings-in-react-tests/

**Comment by Lux-Sales** [comment]
Hey guys! Idk if someone's still having problem with it, but `await waitFor(async () => {})` resolved the warning to me.

**Comment by ontouchstart** [comment]
cat react18-act-environment-explanation.md
# Understanding `window.IS_REACT_ACT_ENVIRONMENT` in React 18 Testing

## Purpose and Function

`window.IS_REACT_ACT_ENVIRONMENT` is a special flag introduced in React 18 that helps manage the warning system for the `act()` testing utility. This flag is a key part of properly testing React applications in concurrent mode.

## How It Works

When set to `true` in your testing environment:

1. **Enables stricter validation**: React will expect that all component updates and state changes happen within `act()` function calls.

2. **Provides better error detection**: You'll get warning messages if state updates occur outside of `act()`, helping identify potential race conditions or async issues.

3. **Enforces proper testing practices**: Ensures that tests properly account for all rendering phases, including deferred or batched updates.

## Why It's Important in React 18

React 18's concurrent rendering introduces new behaviors that make proper testing more challenging:

- **Concurrent rendering**: Updates can be interrupted, deferred, or batched
- **Multiple render passes**: Components may render multiple times before committing
- **Asynchronous behavior**: More operations happen asynchronously compared to React 17

By setting `window.IS_REACT_ACT_ENVIRONMENT = true` in setupTests.js, we're ensuring that our testing environment properly handles these React 18 features.

## Benefits in Testing

1. **Catches race conditions**: Identifies when tests might complete before all rendering phases are done
2. **Improves test reliability**: Ensures tests properly wait for all updates to complete
3. **Reduces flaky tests**: Makes tests more deterministic by properly tracking all rendering phases
4. **Aligns with best practices**: Follows the React team's recommended approach to testing React 18 apps

## Example Scenario

Without this flag, the following test might pass inconsistently:

```javascript
// This might pass sometimes and fail other times without window.IS_REACT_ACT_ENVIRONMENT = true
test('updates counter', () => {
  const { getByText } = render(<Counter />);
  fireEvent.click(getByText('Increment'));
  expect(getByText('Count: 1')).toBeInTheDocument(); // May fail if update hasn't committed
});
```

With the flag enabled, React would warn that the click event might cause updates outside of `act()`, prompting you to use a more reliable testing approach.

**Comment by jakeboone02** [comment]
FYI for anyone using Jest with Ant Design (`antd`) or other `rc-component` libraries, I managed to get Copilot to eliminate these warnings. This was its analysis:

> The issue is that `rc-components` schedules microtasks that fire state updates after the testing library's `act()` wrapper restores `IS_REACT_ACT_ENVIRONMENT` to its previous value. Simply setting `globalThis.IS_REACT_ACT_ENVIRONMENT = true` in setup files doesn't work because Jest runs modules in a sandboxed VM context where Node's `globalThis` isn't visible to `react-dom`.
> 
> The fix is to use Jest's `sandboxInjectedGlobals` config option:
> 
> ```js
> // jest.config.js
> module.exports = {
>   globals: { IS_REACT_ACT_ENVIRONMENT: true },
>   sandboxInjectedGlobals: ['IS_REACT_ACT_ENVIRONMENT'],
>   testEnvironment: 'jsdom',
> };
> ```
> 
> You need both options. The `globals` option sets the value, and `sandboxInjectedGlobals` makes it visible to modules like `react-dom` running in Jest's sandbox.
